File: problems/树/108.将有序数组转换为二叉搜索树.py
#!/usr/bin/env python3
"""
108. 将有序数组转换为二叉搜索树
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sorted_array_to_bst(nums: List[int]) -> Optional[TreeNode]:
    if not nums: return None
    mid = len(nums) // 2
    root = TreeNode(nums[mid])
    root.left  = sorted_array_to_bst(nums[:mid])
    root.right = sorted_array_to_bst(nums[mid+1:])
    return root

def inorder_list(root: Optional[TreeNode]) -> List[int]:
    if not root: return []
    return inorder_list(root.left) + [root.val] + inorder_list(root.right)

if __name__ == "__main__":
    nums1 = [-10, -3, 0, 5, 9]
    root1 = sorted_array_to_bst(nums1)
    print(inorder_list(root1))  # [-10, -3, 0, 5, 9]
    print(inorder_list(sorted_array_to_bst([1, 3])))  # [1, 3]
    print(sorted_array_to_bst([]))  # None

File: problems/树/108.将有序数组转换为二叉搜索树.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* sortedArrayToBST(vector<int>& nums) {
    function<TreeNode*(int,int)> build = [&](int l, int r) -> TreeNode* {
        if (l > r) return nullptr;
        int mid = (l + r) / 2;
        TreeNode* root = new TreeNode(nums[mid]);
        root->left  = build(l, mid-1);
        root->right = build(mid+1, r);
        return root;
    };
    return build(0, (int)nums.size()-1);
}
void inorderPrint(TreeNode* root) {
    if (!root) return;
    inorderPrint(root->left);
    cout << root->val << " ";
    inorderPrint(root->right);
}
int main() {
    vector<int> nums1 = {-10, -3, 0, 5, 9};
    inorderPrint(sortedArrayToBST(nums1)); cout << endl;  // -10 -3 0 5 9
    vector<int> nums2 = {1, 3};
    inorderPrint(sortedArrayToBST(nums2)); cout << endl;  // 1 3
    vector<int> nums3;
    cout << (sortedArrayToBST(nums3) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}
